From 418e1960eef962faa3dda49226a5e5d6057c45f0 Mon Sep 17 00:00:00 2001
From: duli <duli4868@gmail.com>
Date: Thu, 8 Aug 2024 02:25:38 +0800
Subject: [PATCH] feat: avoid latinization of paths in certain functionalities.

Replace `ascii_filename` with `sanitize_file_name` to prevent the
Latinization of paths in library folders and when sending books to
devices.

* src/calibre/db/backend.py (construct_path_name, construct_file_name):
Replace `ascii_filename` with `sanitize_file_name`.

* src/calibre/devices/mtp/driver.py (create_upload_path, upload_books):
Replace `ascii_filename` with `sanitize_file_name`.

* src/calibre/devices/smart_device_app/driver.py (_create_upload_path,
upload_books): Replace `ascii_filename` and its aliases with
`sanitize_file_name`.

* src/calibre/devices/usbms/device.py (sanitize_callback): Replace
`ascii_filename` and its aliases with `sanitize_file_name`.

* src/calibre/devices/usbms/driver.py (upload_books): Add debug
statement.

* src/calibre/gui2/device.py (sync_catalogs, sync_news, sync_to_device):
Replace `ascii_filename` with `sanitize_file_name`. Add debug statement.
---
 src/calibre/db/backend.py                     | 15 ++++++------
 src/calibre/devices/mtp/driver.py             |  5 ++--
 .../devices/smart_device_app/driver.py        | 15 ++++++------
 src/calibre/devices/usbms/device.py           |  5 ++--
 src/calibre/devices/usbms/driver.py           |  1 +
 src/calibre/gui2/device.py                    | 24 ++++++++++++-------
 6 files changed, 37 insertions(+), 28 deletions(-)

diff --git a/src/calibre/db/backend.py b/src/calibre/db/backend.py
index 34977661e1..a593525d9e 100644
--- a/src/calibre/db/backend.py
+++ b/src/calibre/db/backend.py
@@ -21,7 +21,7 @@
 
 import apsw
 
-from calibre import as_unicode, force_unicode, isbytestring, prints
+from calibre import as_unicode, force_unicode, isbytestring, prints, sanitize_file_name
 from calibre.constants import filesystem_encoding, iswindows, plugins, preferred_encoding
 from calibre.db import SPOOL_SIZE, FTSQueryError
 from calibre.db.annotations import annot_db_data, unicode_normalize
@@ -57,7 +57,6 @@
 from calibre.utils.copy_files import copy_files, copy_tree, rename_files, windows_check_if_files_in_use
 from calibre.utils.date import EPOCH, parse_date, utcfromtimestamp, utcnow
 from calibre.utils.filenames import (
-    ascii_filename,
     atomic_rename,
     copyfile_using_links,
     copytree_using_links,
@@ -1471,8 +1470,8 @@ def construct_path_name(self, book_id, title, author):
         '''
         book_id = BOOK_ID_PATH_TEMPLATE.format(book_id)
         l = self.PATH_LIMIT - (len(book_id) // 2) - 2
-        author = ascii_filename(author)[:l]
-        title  = ascii_filename(title.lstrip())[:l].rstrip()
+        author = sanitize_file_name(author)[:l]
+        title  = sanitize_file_name(title.lstrip())[:l].rstrip()
         if not title:
             title = 'Unknown'[:l]
         try:
@@ -1481,7 +1480,7 @@ def construct_path_name(self, book_id, title, author):
         except IndexError:
             author = ''
         if not author:
-            author = ascii_filename(_('Unknown'))
+            author = sanitize_file_name(_('Unknown'))
         if author.upper() in WINDOWS_RESERVED_NAMES:
             author += 'w'
         return f'{author}/{title}{book_id}'
@@ -1498,15 +1497,15 @@ def construct_file_name(self, book_id, title, author, extlen):
         l = (self.PATH_LIMIT - (extlen // 2) - 2) if iswindows else ((self.PATH_LIMIT - extlen - 2) // 2)
         if l < 5:
             raise ValueError('Extension length too long: %d' % extlen)
-        author = ascii_filename(author)[:l]
-        title  = ascii_filename(title.lstrip())[:l].rstrip()
+        author = sanitize_file_name(author)[:l]
+        title  = sanitize_file_name(title.lstrip())[:l].rstrip()
         if not title:
             title = 'Unknown'[:l]
         name   = title + ' - ' + author
         while name.endswith('.'):
             name = name[:-1]
         if not name:
-            name = ascii_filename(_('Unknown'))
+            name = sanitize_file_name(_('Unknown'))
         return name
 
     # Database layer API {{{
diff --git a/src/calibre/devices/mtp/driver.py b/src/calibre/devices/mtp/driver.py
index 532df43b2f..4c36671c01 100644
--- a/src/calibre/devices/mtp/driver.py
+++ b/src/calibre/devices/mtp/driver.py
@@ -410,11 +410,11 @@ def set_plugboards(self, plugboards, pb_func):
 
     def create_upload_path(self, path, mdata, fname, routing):
         from calibre.devices.utils import create_upload_path
-        from calibre.utils.filenames import ascii_filename as sanitize
+        from calibre import sanitize_file_name
         ext = fname.rpartition('.')[-1].lower()
         path = routing.get(ext, path)
 
-        filepath = create_upload_path(mdata, fname, self.save_template, sanitize,
+        filepath = create_upload_path(mdata, fname, self.save_template, sanitize_file_name,
                 prefix_path=path,
                 path_type=posixpath,
                 maxlen=self.MAX_PATH_LEN,
@@ -471,6 +471,7 @@ def upload_books(self, files, names, on_card=None, end_session=True,
 
         for infile, fname, mi in zip(files, names, metadata):
             path = self.create_upload_path(prefix, mi, fname, routing)
+            prints.debug_print("mtp/driver.py:upload_books:path->", path)
             if path and self.is_folder_ignored(storage, path):
                 raise MTPInvalidSendPathError('/'.join(path))
             parent = self.ensure_parent(storage, path)
diff --git a/src/calibre/devices/smart_device_app/driver.py b/src/calibre/devices/smart_device_app/driver.py
index f512914853..059cbca9e6 100644
--- a/src/calibre/devices/smart_device_app/driver.py
+++ b/src/calibre/devices/smart_device_app/driver.py
@@ -21,7 +21,7 @@
 from functools import wraps
 from threading import Thread
 
-from calibre import prints
+from calibre import prints, sanitize_file_name
 from calibre.constants import DEBUG, cache_dir, numeric_version
 from calibre.devices.errors import ControlError, InitialConnectionError, OpenFailed, OpenFeedback, PacketError, TimeoutError, UserFeedback
 from calibre.devices.interface import DevicePlugin, currently_connected_device
@@ -34,9 +34,9 @@
 from calibre.ebooks.metadata.book.base import Metadata
 from calibre.ebooks.metadata.book.json_codec import JsonCodec
 from calibre.library import current_library_name
+from calibre.prints import debug_print
 from calibre.ptempfile import PersistentTemporaryFile
 from calibre.utils.config_base import tweaks
-from calibre.utils.filenames import ascii_filename as sanitize
 from calibre.utils.filenames import shorten_components_to
 from calibre.utils.ipc import eintr_retry_call
 from calibre.utils.mdns import get_all_ips
@@ -447,7 +447,7 @@ def _update_driveinfo_record(self, dinfo, prefix, location_code, name=None):
     # remove the 'path' argument and all its uses. Also removed the calls to
     # filename_callback and sanitize_path_components
     def _create_upload_path(self, mdata, fname, create_dirs=True):
-        fname = sanitize(fname)
+        fname = sanitize_file_name(fname)
         ext = os.path.splitext(fname)[1]
 
         try:
@@ -496,11 +496,11 @@ def _create_upload_path(self, mdata, fname, create_dirs=True):
         id_ = mdata.get('id', fname)
         extra_components = get_components(template, mdata, id_,
                 timefmt=opts.send_timefmt, length=maxlen-len(app_id)-1,
-                last_has_extension=False)
+                last_has_extension=False, sanitize_func=sanitize_file_name)
         if not extra_components:
-            extra_components.append(sanitize(fname))
+            extra_components.append(sanitize_file_name(fname))
         else:
-            extra_components[-1] = sanitize(extra_components[-1]+ext)
+            extra_components[-1] = sanitize_file_name(extra_components[-1]+ext)
 
         if extra_components[-1] and extra_components[-1][0] in ('.', '_'):
             extra_components[-1] = 'x' + extra_components[-1][1:]
@@ -514,7 +514,7 @@ def _create_upload_path(self, mdata, fname, create_dirs=True):
                     extra_components.append('News')
             else:
                 for c in tag.split('/'):
-                    c = sanitize(c)
+                    c = sanitize_file_name(c)
                     if not c:
                         continue
                     extra_components.append(c)
@@ -1488,6 +1488,7 @@ def upload_books(self, files, names, on_card=None, end_session=True,
         for i, infile in enumerate(files):
             mdata, fname = next(metadata), next(names)
             lpath = self._create_upload_path(mdata, fname, create_dirs=False)
+            debug_print("smart_device_app.py/driver.py:upload_books:lpath->",lpath)
             self._debug('lpath', lpath)
             if not hasattr(infile, 'read'):
                 infile = USBMS.normalize_path(infile)
diff --git a/src/calibre/devices/usbms/device.py b/src/calibre/devices/usbms/device.py
index 3285cb5195..681689e3af 100644
--- a/src/calibre/devices/usbms/device.py
+++ b/src/calibre/devices/usbms/device.py
@@ -18,12 +18,11 @@
 from collections import namedtuple
 from itertools import repeat
 
-from calibre import prints
+from calibre import prints, sanitize_file_name
 from calibre.constants import is_debugging, isfreebsd, islinux, ismacos, iswindows
 from calibre.devices.errors import DeviceError
 from calibre.devices.interface import DevicePlugin
 from calibre.devices.usbms.deviceconfig import DeviceConfig
-from calibre.utils.filenames import ascii_filename as sanitize
 from polyglot.builtins import iteritems, string_or_bytes
 
 if ismacos:
@@ -856,7 +855,7 @@ def sanitize_callback(self, path):
         Callback to allow individual device drivers to override the path sanitization
         used by :meth:`create_upload_path`.
         '''
-        return sanitize(path)
+        return sanitize_file_name(path)
 
     def filename_callback(self, default, mi):
         '''
diff --git a/src/calibre/devices/usbms/driver.py b/src/calibre/devices/usbms/driver.py
index a0199b2c52..92002d783d 100644
--- a/src/calibre/devices/usbms/driver.py
+++ b/src/calibre/devices/usbms/driver.py
@@ -327,6 +327,7 @@ def upload_books(self, files, names, on_card=None, end_session=True,
         for i, infile in enumerate(files):
             mdata, fname = next(metadata), next(names)
             filepath = self.normalize_path(self.create_upload_path(path, mdata, fname))
+            debug_print("usbms/driver.py:upload_books:filepath->", filepath)
             if not hasattr(infile, 'read'):
                 infile = self.normalize_path(infile)
             filepath = self.put_file(infile, filepath, replace_file=True)
diff --git a/src/calibre/gui2/device.py b/src/calibre/gui2/device.py
index d8777e430a..05a494d585 100644
--- a/src/calibre/gui2/device.py
+++ b/src/calibre/gui2/device.py
@@ -15,7 +15,7 @@
 from qt.core import QAction, QActionGroup, QCoreApplication, QDialog, QDialogButtonBox, QEventLoop, QIcon, QMenu, QObject, QTimer, QVBoxLayout, pyqtSignal
 
 from calibre import as_unicode, force_unicode, preferred_encoding, prints, sanitize_file_name
-from calibre.constants import DEBUG
+from calibre.constants import DEBUG, is_debugging
 from calibre.customize.ui import available_input_formats, available_output_formats, device_plugins, disabled_device_plugins
 from calibre.devices.errors import (
     BlacklistedDevice,
@@ -53,7 +53,6 @@
 from calibre.ptempfile import force_unicode as filename_to_unicode
 from calibre.startup import connect_lambda
 from calibre.utils.config import device_prefs, tweaks
-from calibre.utils.filenames import ascii_filename
 from calibre.utils.img import scale_image
 from calibre.utils.ipc.job import BaseJob
 from calibre.utils.localization import ngettext
@@ -1428,13 +1427,16 @@ def sync_catalogs(self, send_ids=None, do_auto_convert=True):
             metadata = self.library_view.model().metadata_for(ids)
             names = []
             for book_id, mi in zip(ids, metadata):
-                prefix = ascii_filename(mi.title)
+                prefix = sanitize_file_name(mi.title)
                 if not isinstance(prefix, str):
                     prefix = prefix.decode(preferred_encoding, 'replace')
-                prefix = ascii_filename(prefix)
+                prefix = sanitize_file_name(prefix)
                 names.append('%s_%d%s'%(prefix, book_id,
                     os.path.splitext(files[-1])[1]))
                 self.update_thumbnail(mi)
+            if DEBUG:
+                for name in names:
+                   debug_print("device.py:sync_catalogs: names", name)
             dynamic.set('catalogs_to_be_synced', set())
             if files:
                 remove = []
@@ -1506,13 +1508,16 @@ def sync_news(self, send_ids=None, do_auto_convert=True):
             metadata = self.library_view.model().metadata_for(ids)
             names = []
             for book_id, mi in zip(ids, metadata):
-                prefix = ascii_filename(mi.title)
+                prefix = sanitize_file_name(mi.title)
                 if not isinstance(prefix, str):
                     prefix = prefix.decode(preferred_encoding, 'replace')
-                prefix = ascii_filename(prefix)
+                prefix = sanitize_file_name(prefix)
                 names.append('%s_%d%s'%(prefix, book_id,
                     os.path.splitext(files[-1])[1]))
                 self.update_thumbnail(mi)
+            if DEBUG:
+                for name in names:
+                   debug_print("device.py:sync_news: names", name)
             self.news_to_be_synced = set()
             if config['upload_news_to_device'] and files:
                 remove = ids if del_on_upload else []
@@ -1585,11 +1590,14 @@ def sync_to_device(self, on_card, delete_from_library,
                 a = mi.format_authors()
                 if not a:
                     a = _('Unknown')
-                prefix = ascii_filename(t+' - '+a)
+                prefix = sanitize_file_name(t+' - '+a)
                 if not isinstance(prefix, str):
                     prefix = prefix.decode(preferred_encoding, 'replace')
-                prefix = ascii_filename(prefix)
+                prefix = sanitize_file_name(prefix)
                 names.append('%s_%d%s'%(prefix, id, os.path.splitext(f)[1]))
+        if DEBUG:
+            for name in names:
+                debug_print("device.py:sync_to_device: names", name)
         remove = remove_ids if delete_from_library else []
         self.upload_books(gf, names, good, on_card, memory=(_files, remove))
         self.status_bar.show_message(_('Sending books to device.'), 5000)
-- 
2.46.0

